crates/store/src/filesystem/mod.rs

```diff
@@ -0,0 +1,412 @@
+        key: K,
+        mime: Mime,
+    ) -> Result<Self::Item<'_>, CreateError<Self::Error>> {
+        let (encoded_key, identity) = utility::encode_key(&key).map_err(CreateError::Internal)?;
+
+        if self.get_key_dir(&identity).await.is_ok() {
+            return Err(CreateError::Occupied);
+        }
+
+        self.root
+            .create_dir_all(format!("{}/{}", self.key_root, identity))
+            .map_err(Error::Io)
+            .map_err(CreateError::Internal)?;
+
+        self.write(
+            &format!("{}/{}/{}", self.key_root, identity, KEY_FILE),
+            &encoded_key,
+        )
+        .await
+        .map_err(CreateError::Internal)?;
+
+        Ok(FilesystemCreateItem {
+            fs: self,
+            key_identity: identity,
+            mime,
+            buf: vec![],
+        })
+    }
+}
+
+#[async_trait]
+impl<K> Get<K> for Filesystem<K>
+where
+    K: Serialize + Sync + Send + Eq + Hash,
+{
+    type Item<'a> = Cursor<Vec<u8>> where K: 'a;
+    type Error = Error;
+
+    async fn get(&self, key: K) -> Result<(Meta, Self::Item<'_>), GetError<Self::Error>> {
+        let (_encoded_key, identity) = utility::encode_key(&key).map_err(Error::into_get_error)?;
+        let key_dir = self
+            .get_key_dir(&identity)
+            .await
+            .map_err(Error::into_get_error)?;
+        let mut meta_file = key_dir.clone();
+        meta_file.push(META_FILE);
+        let mut bin_link_file = key_dir.clone();
+        bin_link_file.push(CONTENT_LINK_FILE);
+
+        let meta = self
+            .read_bincode(&meta_file)
+            .await
+            .map_err(|e| Error::BadMetaFile(identity.clone(), Box::new(e)).into_get_error())?;
+        let bin = self
+            .read_bytes(&bin_link_file)
+            .await
+            .map_err(|e| Error::BadBinFile(identity, Box::new(e)).into_get_error())?;
+        Ok((meta, Cursor::new(bin)))
+    }
+}
+
+#[async_trait]
+impl<K> Keys<K> for Filesystem<K>
+where
+    K: DeserializeOwned + 'static + Sync + Send + Clone,
+{
+    type Stream = Iter<std::vec::IntoIter<Result<K, Error>>>;
+    type StreamError = Error;
+
+    async fn keys(&self) -> Self::Stream {
+        iter(self.get_keys().await)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::{
+        fs::{create_dir, remove_dir_all, File},
+        path::Path,
+    };
+
+    use futures::{AsyncReadExt, AsyncWriteExt, StreamExt};
+
+    const TEST_DATA_DIRECTORY: &str = ".test_data";
+
+    #[tokio::test]
+    async fn test() {
+        let key = "test".to_string();
+        let mime = "text/plain".parse::<Mime>().unwrap();
+        let path = Path::new(TEST_DATA_DIRECTORY);
+
+        let _ = remove_dir_all(path);
+        create_dir(path).unwrap();
+        let root_dir = File::open(path).unwrap();
+
+        let mut fs = Filesystem::new(Dir::from_std_file(root_dir.into()), "somestore".to_string());
+
+        assert_eq!(
+            format!(
+                "{:?}",
```

Comment:
Is it intentional that you're comparing the Debug output rather than the Display output?

Reply:
Yeah. Should I use display for this?

Reply:
Perhaps, but I'm also curious as to what's being tested here. Are you testing the output (in which case you should prefer Display), or are you actually testing that the error is the expected type and are trying to avoid deriving PartialEq?

Reply:
using partialeq on Error now :+1: